src/python/pants/engine/target.py

```diff
@@ -106,6 +110,8 @@ def compute_value(cls, raw_value: Optional[Any], *, address: Address) -> Immutab
         The resulting value must be hashable (and should be immutable).
         """
         if raw_value is None:
+            if cls.required:
```

Comment:
As it stands `required` and `default` are disjoint. This is reminiscent of:
```python
@dataclass
class Files(Target):
    sources: Sources
    dependencies: Iterable[str] = None
    tags: Iterable[str] = None
```
Since dataclasses expose metadata (dataclasses.fields(...)), can Target be restructured to leverage more of the language we already have rule authors use?

If not, the smell here is an explosion of Field variant types. You now need a subclass for requiredness, then another for ?... It feels like requiredness should be orthogonal to the type as it is with Optional.

Reply:
> Target be restructured to leverage more of the language we already have rule authors use?

Certainly possible (with some substantial redesign), but this goes against the design goal of a Field-centric API where all the logic lives in Fields and Targets simply pick-and-choose which Fields they want to use.

Contrived, but plausible, example in a world without default globs:

```python
class PythonSources(Sources):
  required = True
  
  def validate_snapshot(snapshot):
     # code to validate only `.py` files

class PythonLibrary(Target):
  alias = "python_library"
  core_fields = (PythonSources, Compatibility, ...)

class PythonTests(Target):
  alias = "python_tests"
  core_fields = (PythonSources, Compatibility, ...)
```

vs.

```python
class PythonSources(Sources): 
  def validate_snapshot(snapshot):
     # code to validate only `.py` files

class PythonLibrary(Target):
  alias = "python_library"
  sources: PythonSources
  compatibility: Optional[Compatiblility]
  ...

class PythonTests(Target):
  alias = "python_tests"
  sources: PythonSources
  compatibility: Optional[Compatiblility]
  ...
```

It's a design goal for a `Target` definition to be no more than 3 lines of code, which means we put logic like "This field is required" in the field itself rather than the target.

--

Likewise, if we make `Fields` be optional on Targets, then we now break the design for most of the core abstractions like `tgt.get(Field)`. 

```python
class Files(Target):
   required_fields = (FilesSources,)
   optional_fields = (Tags,)
```

Now, we can't have MyPy know that `tgt[FilesSources]` returns `FilesSources` rather than `Optional[FilesSources]` because that is runtime information that the field happens to be included in `Files.required_fields`. At the same time, we would need to ensure that `tgt[Tags]` is inferred as `Optional[Tags]`, and then handle the `Optional` in the call site, which is clunky.

--

> If not, the smell here is an explosion of Field variant types. You now need a subclass for requiredness, then another for ?

We already have an explosion of `Fields`. This is the natural consequence of typed fields. For example, we already have `PythonBinarySources`, `PythonLibrarySources`, and `PythonTestsSources`, because they each behave differently.

In writing bindings for the core target types like `Resources` the past two days, I've found that (at least for those bindings), using `required` will only result in one additional Field definition I would have typically avoided: using `ResourcesSources` instead of the parent class `Sources`.

Reply:
Your 1st two examples seem to point to design goals that are arbitrary. In the 1st example I think by lines of code you probably really had as a design goal low boilerplate. Taking a step back and making targets easy to write more generally, the current designs is definitely less lines, but a simple dataclass seems to pull some familiarity weight. In the second example you reference a current API designed to work with the way targets are currently designed. If targets swtiched to plain dataclasses whose fields were Fields, you'd not use `get`, you'd just use the dataclass instance like you do with other python code.

As I understand it, I'm writing a rule and I need 3 fields, say 2 required and 1 optional. So I toss together a lightweight ad-hoc target type (and for now, write a simple adaptor for it). If that target type were a plain old dataclass with Field fields, Optional or not so - I'd interact with it like any other such object, no gets etc. The re-designed Target baseclass could still do all the runtime checks needed via the dataclasses.fields API.

I just want to make sure you're thinking about this with a step back. There may be technical reasons this can't be done. Things were just starting to seem strange from an API user perspective and it wasn't clear to me why that had to be so vs more standard python code.

I'll let you proceed - you have your head in this and can see confounding details all at once that I can't.

Reply:
> design goals that are arbitrary.

Not arbitrary. Not fully explained, sure, because the design goals were generally well established so I didn't relitigate them. The design goals of this project are enumerated in https://docs.google.com/document/d/1nxPdvuzgCPKhTabhfYBN2tbcr85enSR83OpcvdfWXfY/edit#heading=h.6o3hd2sgs0dd.

tl;dr of the relevant bits to this discussion of why we have this Field-driven vs. Target-driven API:
1. Provide downstream rules a structured mechanism for requesting only the Fields they care about and ignoring others.
   * This must work with subclasses of Fields. `PythonSources` can be substituted in for `Sources`, but not vice versa.
2. Extensibility: add new fields to pre-existing target types.
   * We do this now with `UnionRule(PythonLibrary.PluginField, CustomField)`
   * We couldn't easily do this with Targets as dataclasses because Fields would be declared statically at definition time; traditionally, with dataclasses, you'd subclass something to add new attributes to it.
3. Extensibility: create custom target types.
   * You can choose to ignore fields that a spiritual superclass has. E.g. `CustomPythonLibrary` can choose to _not_ register the `Compatibility` field normally associated with `PythonLibrary`. Inheritance of `Targets` (V1 approach) would break this because inheritance makes it difficult/dangerous to deregister an attribute.

For #3, we could say "okay, don't use inheritance" to avoid that, but then we get substantial boilerplate (and still haven't solved #2):

```python
class PythonLibrary(Target):
  alias = "python_library"
  sources: Optional[PythonLibrarySources]
  compatibility: Optional[Compatbility]
  tags: Optional[Tags]
  dependencies: Optional[Dependencies]
   ...
```

vs.

```python
class PythonLibrary(Target):
  alias = "python_library"
  core_fields = (PythonLibrarySources, Compatibility, Tags, Dependencies, ...)
```

Reply:
The explosion of fields problem might be mitigated by the fact that a given field type is likely to always be required or always be optional. I'm struggling to think of an example of a field type that is required in some targets and optional in others. I'm sure there are a few, but maybe it's not common?

Reply:
> The explosion of fields problem might be mitigated by the fact that a given field type is likely to always be required or always be optional. I'm struggling to think of an example of a field type that is required in some targets and optional in others. I'm sure there are a few, but maybe it's not common?

I agree. In practical terms, after having had ported most of the core targets and Python targets, the only case I've found of this is wanting `resources()` to mark `sources` as required, whereas normally we want `sources` to be optional.

The vast majority of the fields have ended up either 

a) not being reused by other target types, e.g. `ZipSafe`, `EntryPoint`, `PrepArgs`, `AliasTargetRequest`
b) being reused but without any modification, e.g. `Dependencies`, `Tags`, and `Compatibility`.

The only field that has seen much subclassing is `Sources`: `PythonSources`, `PythonLibrarySources`, `PythonBinarySources`, `PythonTestsSources`, `FilesSources`, and `ResourcesSources`. Of those, the only that we could avoid if we moved this "required fields" feature from Field definitions to Target definitions is `ResourcesSources`.

Reply:
> I'm struggling to think of an example of a field type that is required in some targets and optional in others

SOunds like this won't be an issue, but your language is the key that make me find this all a bit strange. Its the Target that decides a field is optional or not. Its strange to have a field be inherently required when its only the use case that can ever really say so.

Reply:
> Its strange to have a field be inherently required when its only the use case that can ever really say so.

Well, to be clear, the target author is still making the ultimate decision. They choose whichever fields make sense when declaring their `core_fields`.

If a target author likes some field, minus that it was or was not required, they can easily subclass the original and the API will support that.

For example, we want to create a new `text_files` target that behaves the same as `files()`, including Pants not chrooting the sources, but we want to use default globs of `*.txt` for `sources`. Because we have default globs, we're fine if the user leaves off `sources` in the BUILD file.

```python
class TextSources(FileSources):
  required = False
  default = ("*.txt",)

class TextFiles(Target):
  alias = "text_files"
  core_fields = tuple({*Files.core_fields, TextSources} - {FilesSources})

tgt = TextFiles({}, address=...)
assert tgt[TextSources] == tgt[FilesSources] == tgt[Sources]
```

--

Typed fields are a major paradigm shift away from V1 and a Target-driven API. I can see why it would be weird for a field to determine if it can or cannot be left out of a BUILD file, but that is the spirit of the Target API. 

One of the goals of the project is for all of the logic to move to Fields such that Target definitions do no more than a) register their alias and b) pick which fields they want from the buffet of available fields, including any new fields the author may have written.

This means, for example, that we can put methods on the fields themselves, e.g. this method on `Timeout`:

https://github.com/pantsbuild/pants/blob/e0c08cf7e2180bf2ea6a29bdf646fa5f13b0fd82/src/python/pants/backend/python/rules/targets.py#L128-L140

Conceptually/spiritually, the mindset is "I trust the Field to know what makes sense, e.g. what values it works with, whether it has a default value, whether it has extra methods to do something meaningful with its value", etc. If a Target doesn't like that Field, it can create a new Field.

(I can speak more to _why_ this is important to the project if it'd help. The above comment takes a first pass and https://docs.google.com/document/d/1nxPdvuzgCPKhTabhfYBN2tbcr85enSR83OpcvdfWXfY/edit#heading=h.6o3hd2sgs0dd explains most of it, but it is a major paradigm shift so I'm happy to explain in different words the why.)

--

One other point that came up in discussion with Benjy: to be clear, "optional"/"required" is an overloaded term in this discussion. 

The existence of the field is not optional - if the target type registers `Compatibility`, it will always have that field defined and we will always be able to say `tgt[Compatibility]`

What is optional is whether the user explicitly passes a value in the BUILD file for that field or if we allow them to rely on a default value.

Reply:
It seems like maybe a point we're not connecting on is that there is a BUILD file Target and then the collection of fields a given rule needs. These might be a subset of a BUILD file target and they might be required for the rule but optional for the BUILD file target. For example, a rule that generates javadoc requires a non-empty Sources field. A `java_library` BUILD file target does not.